### PR TITLE
Feature/midx merge tests

### PR DESCRIPTION
**Pull request by ahrav**
## Summary by CodeRabbit

* **Tests**
  * Expanded test coverage for merge/index behavior, large-offset handling and boundary cases, exhaustive deduplication/canonicalization across pack permutations, LOFF round-trip validation, max-duplicate scenarios, and a new "too many objects" limit test.

* **Chores**
  * Updated issue records: set one to in-progress, added an extra test variant to another, and added several closed issues (including a doc-comment task).

### CODE REVIEW

File: crates/scanner-git/src/midx_build.rs [hunk 1]
```diff
@@ -802,6 +1030,169 @@ mod tests {
...
+                                // duplicate resolution remains tied to pack identity rather than
+                                // slice position.
+                                //
+                                // The pack_name clone is required because build_midx_in_memory
+                                // takes &[(u16, Vec<u8>, IdxView)] with owned name bytes.
+                                let views: Vec<_> = permutation
+                                    .iter()
+                                    .map(|&pack_idx| {
+                                        (
+                                            packs[pack_idx].pack_id,
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/scanner-git/src/midx_build.rs]: Good catch — confirmed. I wrote a targeted proof test that builds a 2-pack MIDX with permutation [1, 0] using canonical `pack_id` values, then cross-checks that `pack_id` from OOFF correctly indexes PNAM. It fails exactly as you described: `pack_id=0` resolves to `"pack-1.pack"` instead of `"pack-0.pack"`.

**Root cause**: `build_midx_in_memory` writes PNAM from slice order (line 382) but the test kept canonical `pack_id` values that don't match PNAM position under non-identity permutations. The assertions only compared raw `(pack_id, offset)` tuples, masking the structural mismatch.

**Fix**: The permutation loop now uses `perm_pos as u16` as the pack_id (matching PNAM position), and computes per-permutation expected values via a new `expected_midx_for_order(packs, order)` helper. This ensures the generated MIDX is structurally valid for every permutation. Both the exhaustive test and `all_duplicate_merge_deduplicates_to_lowest_pack_id` were updated.

Note: production code (`build_midx_bytes`) was never affected — it always assigns `pack_id = enumerate position = PNAM position`. The bug was test-only.

File: crates/scanner-git/src/midx_build.rs [hunk 2]
```diff
@@ -802,6 +1030,169 @@ mod tests {
...
+                                            packs[pack_idx].pack_name.clone(),
+                                            idx_views[pack_idx],
+                                        )
+                                    })
+                                    .collect();
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/scanner-git/src/midx_build.rs:1158]: Confirmed — same issue as the kilo-code-bot finding above. I verified with a proof test that `pack_id=0` from OOFF indexes `PNAM[0]` which is `"pack-1.pack"` (wrong) instead of `"pack-0.pack"` under permutation [1, 0].

Fixed by using `perm_pos as u16` as pack_id in both the exhaustive test and the all-duplicates test, with per-permutation expected values computed via `expected_midx_for_order`. The `pack_id` field was also removed from `SmallModelPack` since it's no longer needed.

File: crates/scanner-git/src/midx_build.rs [hunk 3]
```diff
@@ -657,6 +672,219 @@ mod tests {
...
+            oid.copy_from_slice(midx.oid_at(idx));
+            if let Some(prev) = prev_oid {
+                assert!(prev < oid, "OIDL must be strictly ascending at index {idx}");
+            }
+            prev_oid = Some(oid);
+            entries.push((oid, midx.offset_at(idx).unwrap()));
+        }
+
+        CanonicalMidx { fanout, entries }
+    }
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/scanner-git/src/midx_build.rs:891]: Acknowledged — this is accurate. The test helpers (`CanonicalMidx`, `SmallModelPack`, `test_oid`, `canonicalize_midx`) all hardcode `[u8; 20]` for SHA-1.

Since all current tests are explicitly `ObjectFormat::Sha1` and no SHA-256 test paths exist yet, this isn't a current defect. When SHA-256 test coverage is added, these helpers will need to be parameterized over OID length. Deferring for now.

**Comment by ahrav** [comment]
Responding to the three findings from the Greptile review:

---

**1. `merge_entry_ordering` test asserts nothing about `MergeEntry::Ord`** — Confirmed and fixed.

The old test only asserted `oid_a < oid_b` on raw byte arrays without ever touching `MergeEntry`. Replaced it with a real `BinaryHeap<MergeEntry>` test that exercises:
- Lower OID pops first (min-heap ordering via reversed `Ord`)
- Lower `pack_id` wins the tie-break for equal OIDs (the `other.pack_id.cmp(&self.pack_id)` path)

---

**2. `NoPacksFound` variant is never produced** — Confirmed, deferring.

Verified with `grep`: only the definition (`midx_build.rs:59`) and one manually-constructed classification test (`git_executor.rs:769`) reference it. No production code path constructs it. `build_midx_bytes` intentionally succeeds for zero packs.

Since removing the variant touches both `scanner-git` and `gossip-scanner-runtime`, this is better as a standalone cleanup outside this PR's scope.

---

**3. SHA-1 hardcoded `[u8; 20]` in test oracle** — Acknowledged, deferring.

Accurate observation. All current tests are explicitly SHA-1-only. Will rework the test helpers to be parameterized over OID length when SHA-256 test coverage is added.

**Comment by ahrav** [comment]
**Re: CodeRabbit nitpick on LOFF test doc comment (lines 1146-1147)**

Fixed. Reworded from test-justification language ("Verifies the MIDX large-offset path...") to a behavior-focused description:

> Offsets on both sides of the 2^31 LOFF threshold: the serialized MIDX contains both direct OOFF entries and LOFF-indirected 8-byte entries.

File: crates/scanner-git/src/midx_build.rs [hunk 4]
```diff
@@ -689,13 +922,77 @@ mod tests {
 
     #[test]
     fn merge_entry_ordering() {
-        // Create mock entries to test ordering
-        let oid_a = [0x11u8; 20];
-        let oid_b = [0x22u8; 20];
+        // Build a minimal IDX with two OIDs so we can create MergeEntry values
+        // whose Ord impl actually gets exercised.
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/scanner-git/src/midx_build.rs]: Fixed all three. Applied the suggested rewording:

- `merge_entry_ordering`: "Two-OID ordering case: the lower OID must pop before the higher OID."
- `build_midx_loff_boundary_values`: "LOFF threshold boundary: 0x7FFF_FFFF stays inline in OOFF, while 0x8000_0000 is encoded through LOFF indirection."
- `all_duplicate_merge_deduplicates_to_lowest_pack_id`: "Maximum-deduplication case: every pack contains the same OIDs. The merged MIDX contains exactly `objects_per_pack` entries, and each winner resolves to PNAM position 0 for the current input permutation."
